File: src/lib/firebase/repositories/orderRepository.ts
import { collection, query, where, orderBy, onSnapshot, doc, updateDoc, getDoc, Timestamp } from 'firebase/firestore';
import { User } from 'firebase/auth';
import { auth, db } from '@/lib/firebase/config';
import { Order, OrderStatus } from '@/types/order';

export class OrderRepository {
  private static instance: OrderRepository;
  private ordersCollection = collection(db, 'orders');
  
  // Listeners for real-time updates
  private unsubscribeCustomerOrders?: () => void;
  private unsubscribeShopOwnerOrders?: () => void;
  
  // State management
  private customerOrdersCallbacks: ((orders: Order[]) => void)[] = [];
  private shopOwnerOrdersCallbacks: ((orders: Order[]) => void)[] = [];
  
  // Cached data
  private currentCustomerOrders: Order[] = [];
  private currentShopOwnerOrders: Order[] = [];
  
  private constructor() {
    this.setupAuthListener();
  }
  
  static getInstance(): OrderRepository {
    if (!OrderRepository.instance) {
      OrderRepository.instance = new OrderRepository();
    }
    return OrderRepository.instance;
  }
  
  private setupAuthListener() {
    auth.onAuthStateChanged((user: User | null) => {
      if (user) {
        this.startCustomerOrdersListener(user.uid);
        this.startShopOwnerOrdersListener(user.uid);
      } else {
        this.cleanup();
      }
    });
  }
  
  private startCustomerOrdersListener(userId: string) {
    this.cleanup('customer');
    
    // Listen for orders where current user is the customer
    const q = query(
      this.ordersCollection,
      where('customerId', '==', userId)
      // Note: No orderBy to avoid Firestore index requirement, we'll sort in memory
    );
    
    this.unsubscribeCustomerOrders = onSnapshot(q, 
      (snapshot) => {
        console.log('📦 Customer Orders listener update:', snapshot.docs.length, 'orders');
        
        const orders = snapshot.docs
          .map(doc => {
            const data = doc.data();
            return {
              ...data,
              id: doc.id, // Ensure document ID takes precedence over any id field in data
              createdAt: data.createdAt?.toMillis?.() || data.createdAt || 0,
              updatedAt: data.updatedAt?.toMillis?.() || data.updatedAt || 0,
              confirmedAt: data.confirmedAt?.toMillis?.() || data.confirmedAt || 0,
              shippedAt: data.shippedAt?.toMillis?.() || data.shippedAt || 0,
              deliveredAt: data.deliveredAt?.toMillis?.() || data.deliveredAt || 0,
              cancelledAt: data.cancelledAt?.toMillis?.() || data.cancelledAt || 0,
              estimatedDelivery: data.estimatedDelivery?.toMillis?.() || data.estimatedDelivery || 0,
            } as Order;
          })
          .sort((a, b) => b.createdAt - a.createdAt); // Sort by creation date, newest first
        
        this.currentCustomerOrders = orders;
        this.customerOrdersCallbacks.forEach(callback => callback(orders));
      },
      (error) => {
        console.error('❌ Customer Orders listener error:', error);
      }
    );
  }
  
  private startShopOwnerOrdersListener(userId: string) {
    this.cleanup('shopOwner');
    
    // Listen for orders where current user is the shop owner
    const q = query(
      this.ordersCollection,
      where('ownerId', '==', userId)
      // Note: No orderBy to avoid Firestore index requirement, we'll sort in memory
    );
    
    this.unsubscribeShopOwnerOrders = onSnapshot(q, 
      (snapshot) => {
        console.log('📦 Shop Owner Orders listener update:', snapshot.docs.length, 'orders');
        
        const orders = snapshot.docs
          .map(doc => {
            const data = doc.data();
            return {
              ...data,
              id: doc.id, // Ensure document ID takes precedence over any id field in data
              createdAt: data.createdAt?.toMillis?.() || data.createdAt || 0,
              updatedAt: data.updatedAt?.toMillis?.() || data.updatedAt || 0,
              confirmedAt: data.confirmedAt?.toMillis?.() || data.confirmedAt || 0,
              shippedAt: data.shippedAt?.toMillis?.() || data.shippedAt || 0,
              deliveredAt: data.deliveredAt?.toMillis?.() || data.deliveredAt || 0,
              cancelledAt: data.cancelledAt?.toMillis?.() || data.cancelledAt || 0,
              estimatedDelivery: data.estimatedDelivery?.toMillis?.() || data.estimatedDelivery || 0,
            } as Order;
          })
          .sort((a, b) => b.createdAt - a.createdAt); // Sort by creation date, newest first
        
        this.currentShopOwnerOrders = orders;
        this.shopOwnerOrdersCallbacks.forEach(callback => callback(orders));
      },
      (error) => {
        console.error('❌ Shop Owner Orders listener error:', error);
      }
    );
  }
  
  // Subscribe to customer orders (orders where user is the customer)
  subscribeToCustomerOrders(callback: (orders: Order[]) => void): () => void {
    this.customerOrdersCallbacks.push(callback);
    
    // Immediately call with current data if available
    if (this.currentCustomerOrders.length > 0) {
      callback(this.currentCustomerOrders);
    }
    
    return () => {
      this.customerOrdersCallbacks = this.customerOrdersCallbacks.filter(cb => cb !== callback);
    };
  }
  
  // Subscribe to shop owner orders (orders where user owns the shop)
  subscribeToShopOwnerOrders(callback: (orders: Order[]) => void): () => void {
    this.shopOwnerOrdersCallbacks.push(callback);
    
    // Immediately call with current data if available
    if (this.currentShopOwnerOrders.length > 0) {
      console.log('🏪 Repository: Immediately calling new subscriber with', this.currentShopOwnerOrders.length, 'cached orders');
      callback(this.currentShopOwnerOrders);
    }
    
    return () => {
      this.shopOwnerOrdersCallbacks = this.shopOwnerOrdersCallbacks.filter(cb => cb !== callback);
    };
  }
  
  // Update order status
  async updateOrderStatus(orderId: string, status: OrderStatus): Promise<void> {
    try {
      const orderRef = doc(this.ordersCollection, orderId);
      
      const updateData: any = {
        status,
        updatedAt: Timestamp.now()
      };
      
      // Add status-specific timestamps
      switch (status) {
        case OrderStatus.CONFIRMED:
          updateData.confirmedAt = Timestamp.now();
          break;
        case OrderStatus.SHIPPED:
          updateData.shippedAt = Timestamp.now();
          break;
        case OrderStatus.DELIVERED:
          updateData.deliveredAt = Timestamp.now();
          break;
        case OrderStatus.CANCELLED:
          updateData.cancelledAt = Timestamp.now();
          break;
      }
      
      await updateDoc(orderRef, updateData);
      console.log('✅ Order status updated successfully:', orderId, status);
    } catch (error) {
      console.error('❌ Error updating order status:', error);
      throw error;
    }
  }
  
  // Get single order by ID
  async getOrderById(orderId: string): Promise<Order | null> {
    try {
      const orderRef = doc(this.ordersCollection, orderId);
      const docSnap = await getDoc(orderRef);
      
      if (docSnap.exists()) {
        const data = docSnap.data();
        return {
          ...data,
          id: docSnap.id,
          createdAt: data.createdAt?.toMillis?.() || data.createdAt || 0,
          updatedAt: data.updatedAt?.toMillis?.() || data.updatedAt || 0,
          confirmedAt: data.confirmedAt?.toMillis?.() || data.confirmedAt || 0,
          shippedAt: data.shippedAt?.toMillis?.() || data.shippedAt || 0,
          deliveredAt: data.deliveredAt?.toMillis?.() || data.deliveredAt || 0,
          cancelledAt: data.cancelledAt?.toMillis?.() || data.cancelledAt || 0,
          estimatedDelivery: data.estimatedDelivery?.toMillis?.() || data.estimatedDelivery || 0,
        } as Order;
      }
      
      return null;
    } catch (error) {
      console.error('❌ Error fetching order:', error);
      throw error;
    }
  }

  async createOrder(orderData: Omit<Order, 'id' | 'orderNumber'>): Promise<string> {
    try {
      const newOrderRef = doc(this.ordersCollection);
      const orderNumber = `MH-${Date.now()}`;
      
      const finalOrderData = {
        ...orderData,
        id: newOrderRef.id,
        orderNumber,
        createdAt: Timestamp.now(),
        updatedAt: Timestamp.now(),
      };

      await updateDoc(newOrderRef, finalOrderData);
      return newOrderRef.id;
    } catch (error) {
      console.error('❌ Error creating order:', error);
      throw error;
    }
  }
  
  private cleanup(type?: 'customer' | 'shopOwner') {
    if (!type || type === 'customer') {
      if (this.unsubscribeCustomerOrders) {
        this.unsubscribeCustomerOrders();
        this.unsubscribeCustomerOrders = undefined;
      }
      this.currentCustomerOrders = [];
    }
    
    if (!type || type === 'shopOwner') {
      if (this.unsubscribeShopOwnerOrders) {
        this.unsubscribeShopOwnerOrders();
        this.unsubscribeShopOwnerOrders = undefined;
      }
      this.currentShopOwnerOrders = [];
    }
    
    if (!type) {
      this.customerOrdersCallbacks = [];
      this.shopOwnerOrdersCallbacks = [];
    }
  }
}

export const orderRepository = OrderRepository.getInstance();

File: androidAPP/MyHustle/app/src/main/java/com/blueclipse/myhustle/data/repository/OrderRepository.kt
package com.blueclipse.myhustle.data.repository

import android.util.Log
import com.blueclipse.myhustle.data.model.Order
import com.blueclipse.myhustle.data.model.OrderStatus
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.tasks.await

/**
 * Repository for managing order data in Firebase
 */
class OrderRepository private constructor() {
    private val firestore = FirebaseFirestore.getInstance()
    private val ordersCollection = firestore.collection("orders")
    private val auth = FirebaseAuth.getInstance()
    
    // StateFlow for customer orders (orders where current user is the customer)
    private val _orders = MutableStateFlow<List<Order>>(emptyList())
    val orders: StateFlow<List<Order>> = _orders.asStateFlow()
    
    // StateFlow for shop owner orders (orders where current user is the owner)
    private val _shopOwnerOrders = MutableStateFlow<List<Order>>(emptyList())
    val shopOwnerOrders: StateFlow<List<Order>> = _shopOwnerOrders.asStateFlow()
    
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    
    private var listenerErrorCount = 0
    private var lastErrorTime = 0L
    private val maxRetries = 3
    private val errorCooldownMs = 30000L // 30 seconds
    
    // Callback for when orders are created/updated
    private var onOrderCreatedCallback: (() -> Unit)? = null
    
    init {
        // Start listening for current user's orders (as customer)
        startOrdersListener()
        // Start listening for orders where current user is the shop owner
        startShopOwnerOrdersListener()
    }
    
    /**
     * Set callback to be triggered when orders are created or updated
     */
    fun setOnOrderCreatedCallback(callback: () -> Unit) {
        onOrderCreatedCallback = callback
    }
    
    private fun startOrdersListener() {
        val currentUser = auth.currentUser
        if (currentUser != null && shouldRetryListener()) {
            Log.d("OrderRepo", "Starting orders listener for user: ${currentUser.uid}")
            
            // Listen for orders by customer ID WITHOUT orderBy to avoid index requirement
            ordersCollection
                .whereEqualTo("customerId", currentUser.uid)
                .addSnapshotListener { snapshot, error ->
                    if (error != null) {
                        listenerErrorCount++
                        lastErrorTime = System.currentTimeMillis()
                        Log.e("OrderRepo", "Orders listener error (attempt $listenerErrorCount)", error)
                        
                        // Don't restart if we've hit max retries
                        if (listenerErrorCount >= maxRetries) {
                            Log.e("OrderRepo", "Max retries reached, stopping listener to prevent infinite loop")
                            return@addSnapshotListener
                        }
                        return@addSnapshotListener
                    }
                    
                    if (snapshot != null) {
                        // Reset error count on successful response
                        listenerErrorCount = 0
                        
                        Log.d("OrderRepo", "Orders listener update - ${snapshot.documents.size} orders")
                        val userOrders = snapshot.documents.mapNotNull { document ->
                            try {
                                document.toObject(Order::class.java)?.copy(id = document.id)
                            } catch (e: Exception) {
                                Log.e("OrderRepo", "Error parsing order", e)
                                null
                            }
                        }.sortedByDescending { it.createdAt } // Sort in memory instead of Firestore
                        
                        _orders.value = userOrders
                        Log.d("OrderRepo", "Updated orders state with ${userOrders.size} orders")
                    }
                }
        }
        
        // Listen for auth state changes
        auth.addAuthStateListener { firebaseAuth ->
            if (firebaseAuth.currentUser == null) {
                Log.d("OrderRepo", "User logged out, clearing orders")
                _orders.value = emptyList()
                _shopOwnerOrders.value = emptyList()
                resetErrorCounting()
            } else if (firebaseAuth.currentUser != currentUser) {
                Log.d("OrderRepo", "Different user logged in, resetting and restarting listener")
                resetErrorCounting()
                startOrdersListener()
                startShopOwnerOrdersListener()
            }
        }
    }
    
    /**
     * Start real-time listener for orders where current user is the shop owner
     */
    private fun startShopOwnerOrdersListener() {
        val currentUser = auth.currentUser
        if (currentUser != null && shouldRetryListener()) {
            Log.d("OrderRepo", "Starting shop owner orders listener for user: ${currentUser.uid}")
            
            // Listen for orders by owner ID
            ordersCollection
                .whereEqualTo("ownerId", currentUser.uid)
                .addSnapshotListener { snapshot, error ->
                    if (error != null) {
                        listenerErrorCount++
                        lastErrorTime = System.currentTimeMillis()
                        Log.e("OrderRepo", "Shop owner orders listener error (attempt $listenerErrorCount)", error)
                        
                        // Don't restart if we've hit max retries
                        if (listenerErrorCount >= maxRetries) {
                            Log.e("OrderRepo", "Max retries reached, stopping shop owner listener")
                            return@addSnapshotListener
                        }
                        return@addSnapshotListener
                    }
                    
                    if (snapshot != null) {
                        // Reset error count on successful response
                        listenerErrorCount = 0
                        
                        Log.d("OrderRepo", "Shop owner orders listener update - ${snapshot.documents.size} orders")
                        val ownerOrders = snapshot.documents.mapNotNull { document ->
                            try {
                                document.toObject(Order::class.java)?.copy(id = document.id)
                            } catch (e: Exception) {
                                Log.e("OrderRepo", "Error parsing shop owner order", e)
                                null
                            }
                        }.sortedByDescending { it.createdAt }
                        
                        _shopOwnerOrders.value = ownerOrders
                        Log.d("OrderRepo", "Updated shop owner orders state with ${ownerOrders.size} orders")
                    }
                }
        }
    }
    
    private fun shouldRetryListener(): Boolean {
        if (listenerErrorCount >= maxRetries) {
            val timeSinceLastError = System.currentTimeMillis() - lastErrorTime
            if (timeSinceLastError > errorCooldownMs) {
                Log.d("OrderRepo", "Cooldown period elapsed, allowing listener restart")
                resetErrorCounting()
                return true
            }
            Log.d("OrderRepo", "Still in cooldown period, not starting listener")
            return false
        }
        return true
    }
    
    private fun resetErrorCounting() {
        listenerErrorCount = 0
        lastErrorTime = 0L
    }
    
    /**
     * Manually refresh orders from Firestore
     */
    fun refreshOrders() {
        Log.d("OrderRepo", "Manual orders refresh requested")
        if (shouldRetryListener()) {
            startOrdersListener()
        } else {
            Log.w("OrderRepo", "Refresh blocked due to too many errors")
        }
    }
    
    companion object {
        @Volatile
        private var INSTANCE: OrderRepository? = null
        
        val instance: OrderRepository
            get() = INSTANCE ?: synchronized(this) {
                INSTANCE ?: OrderRepository().also { INSTANCE = it }
            }
    }
    
    /**
     * Create a new order
     */
    suspend fun createOrder(order: Order): Result<String> {
        return try {
            _isLoading.value = true
            Log.d("OrderRepo", "Creating order: ${order.orderNumber}")
            val documentRef = ordersCollection.add(order).await()
            val orderId = documentRef.id
            
            // Update the order with its ID
            documentRef.update("id", orderId).await()
            Log.d("OrderRepo", "Order created successfully with ID: $orderId")
            
            // Trigger callback to refresh accounting data
            onOrderCreatedCallback?.invoke()
            
            Result.success(orderId)
        } catch (e: Exception) {
            Log.e("OrderRepo", "Error creating order", e)
            Result.failure(e)
        } finally {
            _isLoading.value = false
        }
    }
    
    /**
     * Get orders for a specific customer
     */
    suspend fun getOrdersForCustomer(customerId: String): Result<List<Order>> {
        return try {
            _isLoading.value = true
            val snapshot = ordersCollection
                .whereEqualTo("customerId", customerId)
                .get()
                .await()
            
            val orders = snapshot.documents.mapNotNull { document ->
                try {
                    document.toObject(Order::class.java)?.copy(id = document.id)
                } catch (e: Exception) {
                    null
                }
            }.sortedByDescending { it.createdAt } // Sort in memory instead
            
            _orders.value = orders
            Result.success(orders)
        } catch (e: Exception) {
            Log.e("OrderRepo", "Error getting orders for customer", e)
            Result.failure(e)
        } finally {
            _isLoading.value = false
        }
    }
    
    /**
     * Get orders for a specific shop owner
     */
    suspend fun getOrdersForShopOwner(ownerId: String): Result<List<Order>> {
        return try {
            _isLoading.value = true
            val snapshot = ordersCollection
                .whereEqualTo("ownerId", ownerId)
                .get()
                .await()
            
            val orders = snapshot.documents.mapNotNull { document ->
                try {
                    document.toObject(Order::class.java)?.copy(id = document.id)
                } catch (e: Exception) {
                    null
                }
            }.sortedByDescending { it.createdAt } // Sort in memory instead
            
            Result.success(orders)
        } catch (e: Exception) {
            Log.e("OrderRepo", "Error getting orders for shop owner", e)
            Result.failure(e)
        } finally {
            _isLoading.value = false
        }
    }
    
    /**
     * Update order status
     */
    suspend fun updateOrderStatus(orderId: String, status: OrderStatus): Result<Unit> {
        return try {
            val updateData = mapOf(
                "status" to status,
                "updatedAt" to System.currentTimeMillis()
            )
            
            when (status) {
                OrderStatus.CONFIRMED -> updateData + ("confirmedAt" to System.currentTimeMillis())
                OrderStatus.SHIPPED -> updateData + ("shippedAt" to System.currentTimeMillis())
                OrderStatus.DELIVERED -> updateData + ("deliveredAt" to System.currentTimeMillis())
                OrderStatus.CANCELLED -> updateData + ("cancelledAt" to System.currentTimeMillis())
                else -> updateData
            }
            
            ordersCollection.document(orderId).update(updateData).await()
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    /**
     * Get order by ID
     */
    suspend fun getOrderById(orderId: String): Result<Order?> {
        return try {
            val document = ordersCollection.document(orderId).get().await()
            if (document.exists()) {
                val order = document.toObject(Order::class.java)?.copy(id = document.id)
                Result.success(order)
            } else {
                Result.success(null)
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    /**
     * Update tracking information
     */
    suspend fun updateTracking(
        orderId: String,
        trackingNumber: String,
        carrier: String
    ): Result<Unit> {
        return try {
            val updateData = mapOf(
                "trackingNumber" to trackingNumber,
                "carrier" to carrier,
                "updatedAt" to System.currentTimeMillis()
            )
            
            ordersCollection.document(orderId).update(updateData).await()
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}
